### ISSUE TITLE
assert_allclose(..., equal_nan=False) doesn't raise an assertion error

### ISSUE THREAD

**Issue by akaihola** [numpy/numpy#8145]
`np.testing.assert_allclose` is supposed to raise an `AssertionError` when comparing two NaN values with `equal_nan=False`. However, it succeeds without exceptions for both scalars and vectors.

`np.allclose(..., equal_nan=False)` does work as documented as seen below, but for `assert_allclose` the behavior is inconsistent with both the documentation and `allclose`.

```
>>> import numpy as np
>>> a = np.array([42.0, np.nan])
>>> b = np.array([42.0, np.nan])
>>> np.allclose(np.nan, np.nan, equal_nan=True)
True
>>> np.allclose(a, b, equal_nan=True)
True
>>> np.testing.assert_allclose(np.nan, np.nan, equal_nan=True)
>>> np.testing.assert_allclose(a, b, equal_nan=True)
>>> np.allclose(np.nan, np.nan, equal_nan=False)
False
>>> np.allclose(a, b, equal_nan=False)
False
>>> # These two should raise an AssertionError:
>>> np.testing.assert_allclose(np.nan, np.nan, equal_nan=False)
>>> np.testing.assert_allclose(a, b, equal_nan=False)
```

I tested this with master branch commit b0a6e55 from today (2016-10-12). I'm running Python 3.5.1 on Fedora 23.

**Comment by akaihola** [comment]
Both `allclose()` and `assert_allclose()` use `numpy.core.numeric.isclose()`, but there's a subtle difference:
- `allclose()` aggregates the results with Python's `all()`
- `assert_allclose()` uses `assert_array_compare()` and passes a curried `isclose()` as the comparison function

`assert_array_compare()` has a hard-coded behavior of always comparing NaNs and infinite values separately, and only applying the comparison function to numeric values. As a result, the `empty_nan=` argument doesn't affect NaN handling at all in `assert_allclose()`.

My proposed fix keeps the behavior of `assert_array_compare()` intact, unless the newly introduced `empty_nan=` argument is set to `False`. In that case it runs both NaNs and numeric values through the comparison function.

**Comment by charris** [comment]
Where is `empty_nan`?

**Comment by akaihola** [comment]
@charris, sorry `empty_nan` was just a typo in the commit message. Should be `equal_nan`.

**Comment by charris** [comment]
Closed by #8165.
